MoreLinq/ToArrayByIndex.cs

```diff
@@ -0,0 +1,241 @@
+        /// projected elements from the input sequence.
+        /// </returns>
+        /// <remarks>
+        /// This method forces immediate query evaluation. It should not be
+        /// used on infinite sequences. If more than one element maps to the
+        /// same index then the latter element overwrites the former in the
+        /// resulting array.
+        /// </remarks>
+
+        public static TResult[] ToArrayByIndex<T, TResult>(this IEnumerable<T> source, int length,
+            Func<T, int> indexSelector, Func<T, TResult> resultSelector)
+        {
+            return source.ToArrayByIndex(length, indexSelector, (e, _) => resultSelector(e));
+        }
+
+        /// <summary>
+        /// Creates an array of user-specified length from an
+        /// <see cref="IEnumerable{T}"/> where a function is used to determine
+        /// the index at which an element will be placed in the array. The
+        /// elements are projected into the array via an additional function.
+        /// </summary>
+        /// <param name="source">The source sequence for the array.</param>
+        /// <param name="length">The length of the resulting array.</param>
+        /// <param name="indexSelector">
+        /// A function that maps an element to its index.</param>
+        /// <param name="resultSelector">
+        /// A function to project a source element into an element of the
+        /// resulting array.</param>
+        /// <typeparam name="T">
+        /// The type of the element in<paramref name="source"/>.</typeparam>
+        /// <typeparam name="TResult">
+        /// The type of the element in the resulting arrary.</typeparam>
+        /// <returns>
+        /// An array of size <paramref name="length"/> that contains the
+        /// projected elements from the input sequence.
+        /// </returns>
+        /// <remarks>
+        /// This method forces immediate query evaluation. It should not be
+        /// used on infinite sequences. If more than one element maps to the
+        /// same index then the latter element overwrites the former in the
+        /// resulting array.
+        /// </remarks>
+
+        public static TResult[] ToArrayByIndex<T, TResult>(this IEnumerable<T> source, int length,
+            Func<T, int> indexSelector, Func<T, int, TResult> resultSelector)
+        {
+            if (source == null) throw new ArgumentNullException("source");
+            if (length < 0) throw new ArgumentOutOfRangeException("length");
+            if (indexSelector == null) throw new ArgumentNullException("indexSelector");
+            if (resultSelector == null) throw new ArgumentNullException("resultSelector");
+
+            var array = new TResult[length];
+            if (length > 0)
+            {
+                foreach (var e in source)
+                {
+                    var i = indexSelector(e);
+                    if (i < array.Length)
```

Comment:
I think this method should throw if the size is not enough. The source list can be filtered by the user before if it may contain values larger than expected, but this method does not know if it is a fatal condition.